pkg/server/commands/batch_check_command.go

```diff
@@ -0,0 +1,182 @@
+	return func(bq *BatchCheckQuery) {
+		bq.maxConcurrentChecks = maxConcurrentChecks
+	}
+}
+
+func WithBatchCheckMaxChecksPerBatch(maxChecks uint32) BatchCheckQueryOption {
+	return func(bq *BatchCheckQuery) {
+		bq.maxChecksAllowed = maxChecks
+	}
+}
+
+func NewBatchCheckCommand(datastore storage.RelationshipTupleReader, checkResolver graph.CheckResolver, typesys *typesystem.TypeSystem, opts ...BatchCheckQueryOption) *BatchCheckQuery {
+	cmd := &BatchCheckQuery{
+		logger:              logger.NewNoopLogger(),
+		datastore:           datastore,
+		cacheController:     cachecontroller.NewNoopCacheController(),
+		checkResolver:       checkResolver,
+		typesys:             typesys,
+		maxChecksAllowed:    config.DefaultMaxChecksPerBatchCheck,
+		maxConcurrentChecks: config.DefaultMaxConcurrentChecksPerBatchCheck,
+	}
+
+	for _, opt := range opts {
+		opt(cmd)
+	}
+	return cmd
+}
+
+func (bq *BatchCheckQuery) Execute(ctx context.Context, params *BatchCheckCommandParams) (map[CorrelationID]*BatchCheckOutcome, *BatchCheckMetadata, error) {
+	if len(params.Checks) > int(bq.maxChecksAllowed) {
+		return nil, nil, errors.ValidationError(
+			fmt.Errorf("batchCheck received %d checks, the maximum allowed is %d ", len(params.Checks), bq.maxChecksAllowed),
+		)
+	}
+
+	if err := validateCorrelationIDs(params.Checks); err != nil {
+		return nil, nil, err
+	}
+
+	var resultMap = new(sync.Map)
+
+	pool := concurrency.NewPool(ctx, int(bq.maxConcurrentChecks))
+	for _, check := range params.Checks {
+		pool.Go(func(ctx context.Context) error {
+			select {
+			case <-ctx.Done():
+				resultMap.Store(check.GetCorrelationId(), &BatchCheckOutcome{
+					Err: context.Canceled,
+				})
+				return nil
+			default:
+			}
+
+			checkQuery := NewCheckCommand(
+				bq.datastore,
+				bq.checkResolver,
+				bq.typesys,
+				WithCheckCommandLogger(bq.logger),
+				WithCacheController(bq.cacheController),
+			)
+
+			checkParams := &CheckCommandParams{
+				StoreID:          params.StoreID,
+				TupleKey:         check.GetTupleKey(),
+				ContextualTuples: check.GetContextualTuples(),
+				Context:          check.GetContext(),
+				Consistency:      params.Consistency,
+			}
+
+			response, _, err := checkQuery.Execute(ctx, checkParams)
+
+			resultMap.Store(check.GetCorrelationId(), &BatchCheckOutcome{
+				CheckResponse: response,
+				Err:           err,
+			})
+
+			return nil
+		})
+	}
+
+	_ = pool.Wait()
+
+	results := map[CorrelationID]*BatchCheckOutcome{}
+	var totalQueryCount uint32
+
+	resultMap.Range(func(k, v interface{}) bool {
+		// Convert types since sync.Map is `any`
+		outcome := v.(*BatchCheckOutcome)
+		results[CorrelationID(k.(string))] = outcome
+
+		totalQueryCount += outcome.CheckResponse.GetResolutionMetadata().DatastoreQueryCount
+		return true
+	})
+
+	return results, &BatchCheckMetadata{TotalQueries: totalQueryCount}, nil
+}
+
+func validateCorrelationIDs(checks []*openfgav1.BatchCheckItem) error {
+	seen := map[string]bool{}
```

Comment:
Slight memory optimization:
```suggestion
	seen := map[string]struct{}{}
```